subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -17,121 +17,148 @@
+                    public Optional<AfterPreviousExecutionState> getAfterPreviousExecutionState() {
+                        return context.getAfterPreviousExecutionState();
                     }
 
                     @Override
-                    public boolean isReused() {
-                        return true;
+                    public Optional<BeforeExecutionState> getBeforeExecutionState() {
+                        return context.getBeforeExecutionState();
                     }
 
                     @Override
-                    public ImmutableSortedMap<String, CurrentFileCollectionFingerprint> getFinalOutputs() {
-                        return finalOutputs;
+                    public UnitOfWork getWork() {
+                        return context.getWork();
                     }
-                };
-            })
-            .orElseGet(() -> {
-                CurrentSnapshotResult executionResult = executeWithoutCache(context);
-                executionResult.getOutcome().ifSuccessfulOrElse(
-                    outcome -> cacheHandler.store(cacheKey -> store(context.getWork(), cacheKey, executionResult)),
-                    failure -> LOGGER.debug("Not storing result of {} in cache because the execution failed", context.getWork().getDisplayName())
-                );
-                return executionResult;
-            });
+                }))
+            )
+            .orElseGet(() -> executeWithCache(cacheHandler, context));
     }
 
-    @Nullable
-    private BuildCacheCommandFactory.LoadMetadata load(UnitOfWork work, BuildCacheKey cacheKey) {
+    private Optional<Try<BuildCacheCommandFactory.LoadMetadata>> load(UnitOfWork work, BuildCacheKey cacheKey) {
         try {
-            return buildCache.load(
-                    commandFactory.createLoad(cacheKey, work, new BuildCacheLoadListener() {
-                        @Override
-                        public void beforeLoad() {
-                            Optional<? extends Iterable<String>> changingOutputs = work.getChangingOutputs();
-                            changingOutputs.ifPresent(affectedPaths -> outputChangeListener.beforeOutputChange(affectedPaths));
-                            if (!changingOutputs.isPresent()) {
-                                outputChangeListener.beforeOutputChange();
-                            }
-                        }
-
-                        @Override
-                        public void afterLoadFailedAndWasCleanedUp(Throwable error) {
-                            work.outputsRemovedAfterFailureToLoadFromCache();
-                        }
-                    })
-            );
-        } catch (UnrecoverableUnpackingException e) {
-            // We didn't manage to recover from the unpacking error, there might be leftover
-            // garbage among the task's outputs, thus we must fail the build
-            throw e;
+            return buildCache.load(commandFactory.createLoad(cacheKey, work))
```

Comment:
This looks a little bit strange. It seems like you actually have a `Try<Optional<LoadMetaData>>` and then want to turn it around by caching the exception yourself instead of using `Try.ofFailable`.

And I think for the `load` method, `Try<Optional<>>` is the right return type, since you always try to load the result.

Reply:
It depends on how you want to handle this. Originally we only handled unpack errors, and this is trying to emulate that behavior, i.e. you can only have a failure if you actually tried to load something.

We could change this, and turn this around (I had an implementation like that before), but I wouldn't do it until we simplify the whole build cache logic in 6.0.

Reply:
IIUC, the code is currently catching any error from the `load` method. We don't know if the failure happened while unpacking or in some other point when loading, right?

The `load` method always tries to load and we capture the failure of loading.

Reply:
You're right, I was thinking of a previous version of the implementation. By turning this logic inside out I actually managed to simplify the code a little. \o/